src/util/datablock/datablock.c

```diff
@@ -72,93 +73,132 @@ void _DataBlock_AddBlocks(DataBlock *dataBlock, size_t blockCount) {
     dataBlock->itemCap = dataBlock->blockCount * BLOCK_CAP;
 }
 
+void static inline _DataBlock_MarkItemAsDeleted(const DataBlock *dataBlock, char* item) {
+    for(int i = 0; i < dataBlock->itemSize; i++) item[i] = DELETED_MARKER;
+}
+
+void static inline _DataBlock_MarkItemAsUndelete(const DataBlock *dataBlock, char* item) {
+    item[0] = !DELETED_MARKER;
+}
+
+int static inline _DataBlock_IsItemDeleted(const DataBlock *dataBlock, char* item) {
+    char deleted = 0xFF;
+    for(int i = 0; i < dataBlock->itemSize; i++) deleted &= (item[i] & DELETED_MARKER);
+    return deleted;
+}
+
+// Checks to see if idx is within global array bounds
+// array bounds are between 0 and itemCount + #deleted indices
+// e.g. [3, 7, 2, D, 1, D, 5] where itemCount = 5 and #deleted indices is 2
+// and so it is valid to query the array with idx 6.
+int static inline _DataBlock_IndexOutOfBounds(const DataBlock *dataBlock, size_t idx) {
+    return (idx < 0 || idx >= (dataBlock->itemCount + array_len(dataBlock->deletedIdx)));
+}
+
 DataBlock *DataBlock_New(size_t itemCap, size_t itemSize) {
-    DataBlock *dataBlock = malloc(sizeof(DataBlock));    
+    DataBlock *dataBlock = malloc(sizeof(DataBlock));
     dataBlock->itemCount = 0;
     dataBlock->itemSize = itemSize;
     dataBlock->blockCount = 0;
     dataBlock->blocks = NULL;
+    dataBlock->deletedIdx = array_new(uint64_t, 128);
     _DataBlock_AddBlocks(dataBlock, ITEM_COUNT_TO_BLOCK_COUNT(itemCap));
     return dataBlock;
 }
 
 DataBlockIterator *DataBlock_Scan(const DataBlock *dataBlock) {
     assert(dataBlock);
     Block *startBlock = dataBlock->blocks[0];
-    return DataBlockIterator_New(startBlock, 0, dataBlock->itemCount, 1);
+
+    int64_t startPos = 0;
+    // Deleted items are skipped, we're about to perform 
+    // array_len(dataBlock->deletedIdx) skips during out scan.
+    int64_t endPos = dataBlock->itemCount + array_len(dataBlock->deletedIdx);
+    return DataBlockIterator_New(startBlock, 0, endPos, 1);
 }
 
-void *DataBlock_GetItem(const DataBlock *dataBlock, size_t idx) {
-    assert(dataBlock && idx >= 0 && idx < dataBlock->itemCount);
-    Block *block = GET_ITEM_BLOCK(dataBlock, idx);
-    idx = ITEM_POSITION_WITHIN_BLOCK(idx);
-    return block->data + (idx * block->itemSize);
+// Make sure datablock can accommodate at least k items.
+void DataBlock_Accommodate(DataBlock *dataBlock, int64_t k) {
+    // Compute number of free slots.
+    int64_t freeSlotsCount = dataBlock->itemCap - dataBlock->itemCount;
+    int64_t additionalItems = k - freeSlotsCount;
+
+    if(additionalItems > 0) {
+        int64_t additionalBlocks = ITEM_COUNT_TO_BLOCK_COUNT(additionalItems);
+        _DataBlock_AddBlocks(dataBlock, additionalBlocks);
+    }
 }
 
-void DataBlock_SetItem(DataBlock *dataBlock, void *item, size_t idx) {
-    assert(dataBlock && idx <= dataBlock->itemCount);
+void *DataBlock_GetItem(const DataBlock *dataBlock, size_t idx) {
+    assert(dataBlock && idx >= 0);
```

Comment:
size_t is unsigned, so second condition is impossible.
We can switch to using `ssize_t` (signed size_t) or `int64_t` in general if you'd like.

Reply:
👍